### PR TITLE
optional type parameter for newPromise()

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

When creating new promises I always find it very frustrating to type the
`: EventLoopPromise<Type>` type annotation but it's necessary for the
compiler to know type the promise will be fulfilled with.

Modifications:

allow an optional `for: SomeType.self` parameter for `newPromise` as

    let p = eventLoop.newPromise(for: Int.self)

is much easier to type than

    let p: EventLoopPromise<Int> = eventLoop.newPromise()

Result:

easier to write code

### CODE REVIEW

**Comment by helje5** [comment]
I know that some people in the server space do this, but I always found those `type.self` things particularly ugly and AFAIK it is not done anywhere else in Swift or Foundation itself. I think the expected API is more like:

    let p = EventLoopPromise<Int>(in: eventLoop)

Which, funny enough, is what `newPromise` is doing. (The `newPromise` thing looks like one of the Java'ish factory patterns you find a lot in NIO.)

**Comment by Lukasa** [comment]
> I know that some people in the server space do this, but I always found those type.self things particularly ugly and AFAIK it is not done anywhere else in Swift or Foundation itself.

Sure it is. Codable uses it heavily, both in the API ([`Decoder.container(keyedBy:)`](https://developer.apple.com/documentation/swift/decoder/2892621-container)) and in the implementation (check out all these [`decode`](https://github.com/apple/swift/blob/806cf57687bb27f15d7ce33426c72d063375a32b/stdlib/public/SDK/Foundation/JSONEncoder.swift#L1359) functions in `JSONEncoder.swift`). It's used in `String` ([`String.decoding(codeUnits:as:)`](https://github.com/apple/swift/blob/806cf57687bb27f15d7ce33426c72d063375a32b/stdlib/public/core/String.swift#L400)), it's used in the unsafe pointer types (see [`UnsafeRawBufferPointer.load(fromByteOffset:as:)`](https://github.com/apple/swift/blob/806cf57687bb27f15d7ce33426c72d063375a32b/stdlib/public/core/UnsafeRawBufferPointer.swift.gyb#L327) and [`UnsafeRawBufferPointer.storeBytes(of:toByteOffset:as:)`](https://github.com/apple/swift/blob/806cf57687bb27f15d7ce33426c72d063375a32b/stdlib/public/core/UnsafeRawBufferPointer.swift.gyb#L361) as examples).

And in fact the last time we did something like this (in 94c21b604375baac0d4f9c7eafde4631a1f1370f)  it was at the suggestion of some members of the standard library team.

TL;DR: the prior art for `.Type` here is pretty good.

The concern about the factory method is reasonable, but as this is intended as a semver minor we can't fix it here. I'm +1 on the idea of changing to a construct for SwiftNIO 2.0 though.

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -318,7 +318,7 @@ extension EventLoop {
     }
 
     /// Creates and returns a new `EventLoopPromise` that will be notified using this `EventLoop` as execution `Thread`.
-    public func newPromise<T>(file: StaticString = #file, line: UInt = #line) -> EventLoopPromise<T> {
+    public func newPromise<T>(type: T.Type = T.self, file: StaticString = #file, line: UInt = #line) -> EventLoopPromise<T> {
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: I don't like the name `type` here. It sits weirdly with me: it's not the type of the promise, it's the type of what it contains. Not sure what the better name choice is though: `resultType`, maybe? Or `futureResult`? Or `producing`?

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: @Lukasa `makePromise<T>(for type: T.Type = T.self)` ? that reads okay: 'make promise for integer'.

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: well, in 1.0 obviously `newPromise`

Comment (Lukasa) [reply Sources/NIO/EventLoop.swift]: Works for me.

**Comment by weissi** [comment]
@Lukasa / @normanmaurer I'm open to opening up `EventLoopPromise`'s `init` and making it public but I'm against removing the factory from the event loop. I think we can have both.

**Comment by Lukasa** [comment]
I'm happy enough with that as a compromise position.

File: Sources/NIO/EventLoop.swift (hunk 1 above)
Comment (weissi) [reply Sources/NIO/EventLoop.swift]: @Lukasa updated

Comment (ianpartridge) [reply Sources/NIO/EventLoop.swift]: Would `makePromise<T>(of type: T.Type = T.self)` be better?

Comment (ktoso) [reply Sources/NIO/EventLoop.swift]: Seems I'm also late to the party here, and wanted to suggest `of` as well when I was reading the PR. "a promise of string (in the future)" somehow reads better than "a promise for string (in the future)". Your call tho

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: we can still change it, it's not released. Let's page stdlib folk for API guidance, CC @airspeedswift / @lorentey

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: ok, we now landed on `newPromise(of: T.Type = T.self)` because 'making a promise of a type' is more grammatical than 'making a promise for a type' -->  #672
